File: VB/ServiceDataStore/CachingServiceDataStore.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.Xpo.DB
Imports System.Web.Services
Imports DevExpress.Xpo.DB.Helpers
Imports System.Web.Services.Protocols

Namespace DXSample.Proxy
	<WebServiceBinding(Namespace := "http://devexpress.example/")> _
	Public Class CachingDataStore
		Inherits SoapHttpClientProtocol
		Implements ICachedDataStore
		Public Sub New(ByVal url As String)
			MyBase.New()
			Me.Url = url
		End Sub

		<SoapDocumentMethod("http://devexpress.example/ModifyData", RequestNamespace := "http://devexpress.example/", ResponseNamespace := "http://devexpress.example/")> _
		Public Overridable Function ModifyData(ByVal cookie As DataCacheCookie, ByVal dmlStatements() As ModificationStatement) As DataCacheModificationResult
			Return CType(Invoke("ModifyData", New Object() { cookie, dmlStatements })(0), DataCacheModificationResult)
		End Function

		<SoapDocumentMethod("http://devexpress.example/NotifyDirtyTables", RequestNamespace := "http://devexpress.example/", ResponseNamespace := "http://devexpress.example/")> _
		Public Overridable Function NotifyDirtyTables(ByVal cookie As DataCacheCookie, ParamArray ByVal dirtyTablesNames() As String) As DataCacheResult
			Return CType(Invoke("NotifyDirtyTables", New Object() { cookie, dirtyTablesNames })(0), DataCacheResult)
		End Function

		<SoapDocumentMethod("http://devexpress.example/ProcessCookie", RequestNamespace := "http://devexpress.example/", ResponseNamespace := "http://devexpress.example/")> _
		Public Overridable Function ProcessCookie(ByVal cookie As DataCacheCookie) As DataCacheResult
			Return CType(Invoke("ProcessCookie", New Object() { cookie })(0), DataCacheResult)
		End Function

		<SoapDocumentMethod("http://devexpress.example/SelectData", RequestNamespace := "http://devexpress.example/", ResponseNamespace := "http://devexpress.example/")> _
		Public Overridable Function SelectData(ByVal cookie As DataCacheCookie, ByVal selects() As SelectStatement) As DataCacheSelectDataResult
			Return CType(Invoke("SelectData", New Object() { cookie, selects })(0), DataCacheSelectDataResult)
		End Function

		Protected Overridable Function UpdateSchema(ByVal cookie As DataCacheCookie, ByVal tables() As DBTable, ByVal dontCreateIfFirstTableNotExits As Boolean) As DataCacheUpdateSchemaResult
			Throw New InvalidOperationException("Schema modification is not allowed")
		End Function

		#Region "ICacheToCacheCommunicationCore Members"

		Private Function ICacheToCacheCommunicationCore_ModifyData(ByVal cookie As DataCacheCookie, ByVal dmlStatements() As ModificationStatement) As DataCacheModificationResult Implements ICacheToCacheCommunicationCore.ModifyData
			Return ModifyData(cookie, dmlStatements)
		End Function

		Private Function ICacheToCacheCommunicationCore_NotifyDirtyTables(ByVal cookie As DataCacheCookie, ParamArray ByVal dirtyTablesNames() As String) As DataCacheResult Implements ICacheToCacheCommunicationCore.NotifyDirtyTables
			Return NotifyDirtyTables(cookie, dirtyTablesNames)
		End Function

		Private Function ICacheToCacheCommunicationCore_ProcessCookie(ByVal cookie As DataCacheCookie) As DataCacheResult Implements ICacheToCacheCommunicationCore.ProcessCookie
			Return ProcessCookie(cookie)
		End Function

		Private Function ICacheToCacheCommunicationCore_SelectData(ByVal cookie As DataCacheCookie, ByVal selects() As SelectStatement) As DataCacheSelectDataResult Implements ICacheToCacheCommunicationCore.SelectData
			Return SelectData(cookie, selects)
		End Function

		Private Function ICacheToCacheCommunicationCore_UpdateSchema(ByVal cookie As DataCacheCookie, ByVal tables() As DBTable, ByVal dontCreateIfFirstTableNotExist As Boolean) As DataCacheUpdateSchemaResult Implements ICacheToCacheCommunicationCore.UpdateSchema
			Return UpdateSchema(cookie, tables, dontCreateIfFirstTableNotExist)
		End Function

		#End Region

		#Region "IDataStore Members"

		Private ReadOnly Property IDataStore_AutoCreateOption() As AutoCreateOption Implements IDataStore.AutoCreateOption
			Get
				Return AutoCreateOption.SchemaAlreadyExists
			End Get
		End Property

		Private Function ModifyData(ParamArray ByVal dmlStatements() As ModificationStatement) As ModificationResult Implements IDataStore.ModifyData
			Return ModifyData(DataCacheCookie.Empty, dmlStatements).ModificationResult
		End Function

		Private Function SelectData(ParamArray ByVal selects() As SelectStatement) As SelectedData Implements IDataStore.SelectData
			Return SelectData(DataCacheCookie.Empty, selects).SelectedData
		End Function

		Private Function IDataStore_UpdateSchema(ByVal dontCreateIfFirstTableNotExist As Boolean, ParamArray ByVal tables() As DBTable) As UpdateSchemaResult Implements IDataStore.UpdateSchema
			Return UpdateSchema(DataCacheCookie.Empty, tables, dontCreateIfFirstTableNotExist).UpdateSchemaResult
		End Function

		#End Region
	End Class
End Namespace

File: CS/ServiceDataStore/CachingServiceDataStore.cs
using System;
using DevExpress.Xpo.DB;
using System.Web.Services;
using DevExpress.Xpo.DB.Helpers;
using System.Web.Services.Protocols;

namespace DXSample.Proxy {
    [WebServiceBinding(Namespace = "http://devexpress.example/")]
    public class CachingDataStore : SoapHttpClientProtocol, ICachedDataStore {
        public CachingDataStore(string url)
            : base() {
            Url = url;
        }

        [SoapDocumentMethod("http://devexpress.example/ModifyData",
            RequestNamespace = "http://devexpress.example/",
            ResponseNamespace = "http://devexpress.example/")]
        public virtual DataCacheModificationResult ModifyData(DataCacheCookie cookie,
            ModificationStatement[] dmlStatements) {
            return (DataCacheModificationResult)Invoke("ModifyData",
                new object[] { cookie, dmlStatements })[0];
        }

        [SoapDocumentMethod("http://devexpress.example/NotifyDirtyTables",
            RequestNamespace = "http://devexpress.example/",
            ResponseNamespace = "http://devexpress.example/")]
        public virtual DataCacheResult NotifyDirtyTables(DataCacheCookie cookie,
            params string[] dirtyTablesNames) {
            return (DataCacheResult)Invoke("NotifyDirtyTables",
                new object[] { cookie, dirtyTablesNames })[0];
        }

        [SoapDocumentMethod("http://devexpress.example/ProcessCookie",
            RequestNamespace = "http://devexpress.example/",
            ResponseNamespace = "http://devexpress.example/")]
        public virtual DataCacheResult ProcessCookie(DataCacheCookie cookie) {
            return (DataCacheResult)Invoke("ProcessCookie", new object[] { cookie })[0];
        }

        [SoapDocumentMethod("http://devexpress.example/SelectData",
            RequestNamespace = "http://devexpress.example/",
            ResponseNamespace = "http://devexpress.example/")]
        public virtual DataCacheSelectDataResult SelectData(DataCacheCookie cookie,
            SelectStatement[] selects) {
            return (DataCacheSelectDataResult)Invoke("SelectData", new object[] { cookie, selects })[0];
        }

        protected virtual DataCacheUpdateSchemaResult UpdateSchema(DataCacheCookie cookie,
            DBTable[] tables, bool dontCreateIfFirstTableNotExits) {
            throw new InvalidOperationException("Schema modification is not allowed");
        }

        #region ICacheToCacheCommunicationCore Members

        DataCacheModificationResult ICacheToCacheCommunicationCore.ModifyData(DataCacheCookie cookie,
            ModificationStatement[] dmlStatements) {
            return ModifyData(cookie, dmlStatements);
        }

        DataCacheResult ICacheToCacheCommunicationCore.NotifyDirtyTables(DataCacheCookie cookie, 
            params string[] dirtyTablesNames) {
            return NotifyDirtyTables(cookie, dirtyTablesNames);
        }

        DataCacheResult ICacheToCacheCommunicationCore.ProcessCookie(DataCacheCookie cookie) {
            return ProcessCookie(cookie);
        }

        DataCacheSelectDataResult ICacheToCacheCommunicationCore.SelectData(DataCacheCookie cookie,
            SelectStatement[] selects) {
            return SelectData(cookie, selects);
        }

        DataCacheUpdateSchemaResult ICacheToCacheCommunicationCore.UpdateSchema(DataCacheCookie cookie,
            DBTable[] tables, bool dontCreateIfFirstTableNotExist) {
            return UpdateSchema(cookie, tables, dontCreateIfFirstTableNotExist);
        }

        #endregion

        #region IDataStore Members

        AutoCreateOption IDataStore.AutoCreateOption {
            get { return AutoCreateOption.SchemaAlreadyExists; }
        }

        ModificationResult IDataStore.ModifyData(params ModificationStatement[] dmlStatements) {
            return ModifyData(DataCacheCookie.Empty, dmlStatements).ModificationResult;
        }

        SelectedData IDataStore.SelectData(params SelectStatement[] selects) {
            return SelectData(DataCacheCookie.Empty, selects).SelectedData;
        }

        UpdateSchemaResult IDataStore.UpdateSchema(bool dontCreateIfFirstTableNotExist, 
            params DBTable[] tables) {
            return UpdateSchema(DataCacheCookie.Empty, tables, 
                dontCreateIfFirstTableNotExist).UpdateSchemaResult;
        }

        #endregion
    }
}